File: top_scores.rb
def top_scores(scores, highest_possible)
  # Counting sort: O(n)
  counts = Array.new(highest_possible, 0)

  # Count number of each score
  for score in scores
    counts[score] += 1
  end

  # Aggregate previous count to each count
  (1..counts.length-1).each do |i|
    counts[i] += counts[i - 1]
  end

  sorted_scores = Array.new(scores.length) 
  for score in scores
    sorted_scores[counts[score] - 1] = score
    counts[score] -= 1
  end 

  return sorted_scores.inspect
end

scores = [99, 8, 1, 3, 54, 34, 24, 2, 4, 2, 2]
puts top_scores(scores, 100)

File: top_scores.py
# Given a list of unsorted scores and the highest possible score,
# return a sorted list of scores in less than O(n lg n) time.

# counting sort
# T: O(n + k)
# S: O(n + k)
def sort_scores(unsorted_scores, highest_score):
    counts = [0] * (highest_score + 1)
    output = []

    for score in unsorted_scores:
        counts[score] += 1

    # counts = [1, 1, 2]
    for score in xrange(highest_score, -1, -1):
        for time in xrange(counts[score]):
            output.append(score)

    return output

print sort_scores([3, 3, 1, 2], 3) # [1, 2, 3, 3]
print sort_scores([37, 89, 41, 65, 91, 53], 100) # [91, 89, 65, 53, 41, 37]
